### PR TITLE
many: drop NameAndRevision, use snap.PlaceInfo instead

### PR DESCRIPTION
**Pull request by anonymouse64**
* Add a `SnapRevision()` method to `snap.PlaceInfo` so that it can replace `boot.NameAndRevision`
* Add `ParsePlaceInfoFromSnapFileName` (yes very long and specific name, open to alternatives) which was adapted from `nameAndRevnoFromSnap` and expanded upon to do a little bit more error checking and added more unit test cases around it.
* Remove `boot.NameAndRevision` and use `snap.PlaceInfo` instead and remove `nameAndRevnoFromSnap` and use `snap.ParsePlaceInfoFromSnapFileName` everywhere
* Tweaked `revisions()` in `bootState16` a bit to check if the variable is empty and emit the same existing specific error message there, which reads better than the error message we would have gotten from `snap.ParsePlaceInfoFromSnapFileName` without this check:
```
cannot get name and revision of kernel (snap_kernel): empty snap file name
```

This is stacked on top of https://github.com/snapcore/snapd/pull/7973 and https://github.com/snapcore/snapd/pull/7976 since there are some conflicts between these changes and those PR's.

The relevant commits are:

- https://github.com/snapcore/snapd/commit/406cd33795408ff99050cb4d92c2d340903ee3f4
- https://github.com/snapcore/snapd/commit/746b2b7787400dd9e99ee08196161ff9e46bc681
- https://github.com/snapcore/snapd/commit/9e8fc9c5b18fa032d9dd8e2c563b0882ed0efb52

### CODE REVIEW

File: snap/info.go [hunk 1]
```diff
@@ -89,6 +89,35 @@ func MinimalPlaceInfo(name string, revision Revision) PlaceInfo {
...
 
+// ParsePlaceInfoFromSnapFileName returns a PlaceInfo with just the location
+// information for a snap of file name, failing if the snap file name is invalid
+// This explicitly does not support filenames with instance names in them
+func ParsePlaceInfoFromSnapFileName(sn string) (PlaceInfo, error) {
+	if sn == "" {
+		return nil, fmt.Errorf("empty snap file name")
+	}
+	if strings.Count(sn, "_") > 1 {
+		// too many "_", probabl has an instance key in the filename like in
```
Comment (pedronis) [inline snap/info.go]: typo: probably

File: snap/info.go [hunk 2]
```diff
@@ -89,6 +89,35 @@ func MinimalPlaceInfo(name string, revision Revision) PlaceInfo {
...
+		return nil, fmt.Errorf("snap file name %q has invalid format (missing '_')", sn)
+	case idx == 0:
+		return nil, fmt.Errorf("snap file name %q has invalid format (no snap name before '_')", sn)
+	}
+	// ensure that _ is not the last element
+	name := sn[:idx]
+	revnoNSuffix := sn[idx+1:]
+	rev, err := ParseRevision(strings.TrimSuffix(revnoNSuffix, ".snap"))
+	if err != nil {
+		return nil, err
```
Comment (pedronis) [inline snap/info.go]: maybe prefix with  "cannot parse revision in snap file name %q: ... " ?

Comment (anonymouse64) [reply snap/info.go]: sure, that makes sense

**Comment by anonymouse64** [comment]
Preparing uc20 here timed out again:

```
2020-01-09 21:50:45 Error preparing google:ubuntu-core-20-64:tests/core20/ : kill-timeout reached after jan092110-393673 (google:ubuntu-core-20-64:tests/core20/) reboot request
travis_time:end:17307ea6:start=1578604733133782694,finish=1578606645643277246,duration=1912509494552
travis_time:start:61a3fc16
2020-01-09 21:50:45 Restoring google:ubuntu-core-20-64:tests/core20/...
travis_time:end:61a3fc16:start=1578606645643312150,finish=1578606645643362231,duration=50081
travis_time:start:61a3fc17
2020-01-09 21:50:45 Error restoring google:ubuntu-core-20-64:tests/core20/ : EOF
travis_time:end:61a3fc17:start=1578606645643312151,finish=1578606645643374926,duration=62775
travis_time:start:61a50f3b
2020-01-09 21:50:45 Restoring google:ubuntu-core-20-64...
travis_time:end:61a50f3b:start=1578606645643382587,finish=1578606645643404842,duration=22255
travis_time:start:61a50f3c
2020-01-09 21:50:45 Error restoring google:ubuntu-core-20-64 : EOF
travis_time:end:61a50f3c:start=1578606645643382588,finish=1578606645643420093,duration=37505
2020-01-09 21:50:45 Discarding google:ubuntu-core-20-64 (jan092110-393673)...
travis_time:end:47ffc00b:start=1578604733952671755,finish=1578606650713733063,duration=1916761061308
travis_time:start:47ffc00c
2020-01-09 21:50:50 Error preparing google:ubuntu-core-20-64:tests/smoke/ : kill-timeout reached after jan092110-393682 (google:ubuntu-core-20-64:tests/smoke/) reboot request
travis_time:end:47ffc00c:start=1578604733952671756,finish=1578606650713778436,duration=1916761106680
travis_time:start:8fddd22c
2020-01-09 21:50:50 Restoring google:ubuntu-core-20-64:tests/smoke/...
travis_time:end:8fddd22c:start=1578606650713821740,finish=1578606650713865964,duration=44224
travis_time:start:8fddd22d
2020-01-09 21:50:50 Error restoring google:ubuntu-core-20-64:tests/smoke/ : read tcp 10.20.1.153:47792->34.73.226.250:22: read: connection reset by peer
travis_time:end:8fddd22d:start=1578606650713821741,finish=1578606650713883985,duration=62244
travis_time:start:8fdee9df
2020-01-09 21:50:50 Restoring google:ubuntu-core-20-64...
travis_time:end:8fdee9df:start=1578606650713893343,finish=1578606650713911295,duration=17952
travis_time:start:8fdee9e0
2020-01-09 21:50:50 Error restoring google:ubuntu-core-20-64 : read tcp 10.20.1.153:47792->34.73.226.250:22: read: connection reset by peer
travis_time:end:8fdee9e0:start=1578606650713893344,finish=1578606650713921782,duration=28438
2020-01-09 21:50:50 Discarding google:ubuntu-core-20-64 (jan092110-393682)...
travis_time:end:14642f44:start=1578604741676773188,finish=1578606651701958439,duration=1910025185251
travis_time:start:14642f45
2020-01-09 21:50:51 Error preparing google:ubuntu-core-20-64:tests/regression/ : kill-timeout reached after jan092110-393665 (google:ubuntu-core-20-64:tests/regression/) reboot request
travis_time:end:14642f45:start=1578604741676773189,finish=1578606651702008902,duration=1910025235713
travis_time:start:cac4c823
2020-01-09 21:50:51 Restoring google:ubuntu-core-20-64:tests/regression/...
travis_time:end:cac4c823:start=1578606651702036515,finish=1578606651702070985,duration=34470
travis_time:start:cac4c824
2020-01-09 21:50:51 Error restoring google:ubuntu-core-20-64:tests/regression/ : EOF
travis_time:end:cac4c824:start=1578606651702036516,finish=1578606651702082634,duration=46118
travis_time:start:cac59b93
2020-01-09 21:50:51 Restoring google:ubuntu-core-20-64...
travis_time:end:cac59b93:start=1578606651702090643,finish=1578606651702108688,duration=18045
travis_time:start:cac59b94
2020-01-09 21:50:51 Error restoring google:ubuntu-core-20-64 : EOF
travis_time:end:cac59b94:start=1578606651702090644,finish=1578606651702115427,duration=24783
2020-01-09 21:50:51 Discarding google:ubuntu-core-20-64 (jan092110-393665)...
2020-01-09 21:50:52 Successful tasks: 3341
2020-01-09 21:50:52 Aborted tasks: 280
2020-01-09 21:50:52 Failed suite prepare: 6
    - google:ubuntu-core-20-64:tests/core20/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/regression/
    - google:ubuntu-core-20-64:tests/regression/
    - google:ubuntu-core-20-64:tests/smoke/
2020-01-09 21:50:52 Failed suite restore: 9
    - google:ubuntu-core-20-64:tests/core20/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/main/
    - google:ubuntu-core-20-64:tests/regression/
    - google:ubuntu-core-20-64:tests/regression/
    - google:ubuntu-core-20-64:tests/regression/
    - google:ubuntu-core-20-64:tests/smoke/
2020-01-09 21:50:52 Failed project restore: 6
    - google:ubuntu-core-20-64:project
    - google:ubuntu-core-20-64:project
    - google:ubuntu-core-20-64:project
    - google:ubuntu-core-20-64:project
    - google:ubuntu-core-20-64:project
    - google:ubuntu-core-20-64:project
error: unsuccessful run
```

**Pull request merged**